src/BackgroundTask/SingleScheduledBackgroundTask.php

```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * Abstract class SingleScheduledBackgroundTask.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\BackgroundTask;
+
+use AmpProject\AmpWP\Infrastructure\Conditional;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+
+/**
+ * Abstract base class for using cron to execute a background task that runs only once.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.1
+ * @internal
+ */
+abstract class SingleScheduledBackgroundTask implements Service, Registerable, Conditional {
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		return is_admin() || wp_doing_cron();
+	}
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param BackgroundTaskDeactivator $background_task_deactivator Service that deactivates background events.
+	 */
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator ) {
+		$background_task_deactivator->add_event( $this->get_event_name() );
```

Comment:
Wouldn't this make more sense to be called in the `register` method?

Reply:
Or, come to think of it, there could be a default `register` implementation that does `$this->background_task_deactivator->add_event( $this->get_event_name() )`. The constructor could be just responsible for storing `$background_task_deactivator` as a member variable.